backend/ethereum/channel/reorg_test.go

```diff
@@ -0,0 +1,180 @@
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove Approval.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return []types.Transactions{nil, nil, txs[1]}
+		}))
+		// Both TX invalid.
+		s.ConfirmTx(tx1, false)
+		s.ConfirmTx(tx2, false)
+		// All events removed.
+		s.AllowanceEvent(1, false)
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false)
+		s.NoMoreEvents()
+	})
+	t.Run("remove-transfer", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove transfer.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return []types.Transactions{txs[0]}
+		}))
+		// Wait for approval, but not transfer.
+		s.ConfirmTx(tx1, true)
+		s.ConfirmTx(tx2, false)
+		// Check that the transfer rollbacked.
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false) // allowance is 1 again
+		s.NoMoreEvents()
+	})
+	t.Run("remove-transfer-rebirth", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove transfer.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return []types.Transactions{nil, nil, txs[0]}
+		}))
+		// Wait for approval, but not transfer.
+		s.ConfirmTx(tx1, true)
+		s.ConfirmTx(tx2, false)
+		// Check that the transfer rollbacked.
+		s.AllowanceEvent(1, false) // allowance is 0
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false)
+		// Rebirth
+		s.AllowanceEvent(1, true) // allowance is 1 again
+		s.NoMoreEvents()
+	})
+	t.Run("remove-both", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove transfer.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return nil
+		}))
+		// Both TX invalid.
+		s.ConfirmTx(tx1, false)
+		s.ConfirmTx(tx2, false)
+		// All events removed.
+		s.AllowanceEvent(1, false)
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false)
+		s.NoMoreEvents()
+	})
+	t.Run("switch-tx", func(t *testing.T) {
+		rng := pkgtest.Prng(t)
+		s := test.NewTokenSetup(ctx, t, rng)
+		s.StartSubs()
+		defer s.CloseSubs()
+
+		// Send TXs.
+		tx1, tx2 := sendTXs(ctx, s)
+		// Reorg, remove transfer.
+		require.NoError(t, s.SB.Reorg(ctx, 2, 3, func(txs []types.Transactions) []types.Transactions {
+			return []types.Transactions{txs[1], txs[0]}
+		}))
+		// Wait for approval, but not transfer.
+		s.ConfirmTx(tx1, true)
+		s.ConfirmTx(tx2, false)
+		// All events removed.
+		s.AllowanceEvent(1, false)
+		s.TransferEvent(false)
+		s.AllowanceEvent(0, false)
+		// Allowance event reborn.
+		s.AllowanceEvent(1, true)
```

Comment:
Again, can you explain what happens here? Why is it true now?

Reply:
It is similar to what happens in Z116.  
If an event gets re-orged to a different block, it will be reborn (aka. re-emitted).